File: Questions/range_sum_queries/range_sum_queries.py
from itertools import accumulate
class DSA:
    def range_sum_queries(self, m, n, arr, queries):
        acc = tuple(accumulate(arr))
        res = [0] * n
        for idx, query in enumerate(queries):
            if query[0] == 0:
                res[idx] += acc[query[1]]
            else:
                res[idx] += (acc[query[1]] - acc[query[0]-1])
        return res

dsa = DSA()
m, n = map(int, input().split())
arr = list(map(int, input().split()))
queries = [tuple(map(int, input().split())) for _ in range(n)]
res = dsa.range_sum_queries(m, n, arr, queries)
print(arr)
for i in range(n):
    print(queries[i], res[i])

File: Topic-wise/Introduction/range_sum_queries/range_sum_queries.rs
use std::io;

struct DSA;

impl DSA {
    pub fn range_sum_queries(
        m: usize,
        n: usize,
        arr: Vec<i32>,
        queries: Vec<(i32, i32)>,
    ) -> Vec<i32> {
        let acc: Vec<i32> = arr
            .iter()
            .scan(0, |state, x| {
                *state += x;
                Some(*state)
            })
            .collect();

        let mut res = vec![0; queries.len()];
        for i in 0..n {
            if queries[i as usize].0 == 0 {
                res[i] += acc[queries[i].1 as usize];
            } else {
                res[i] +=
                    acc[queries[i as usize].1 as usize] - acc[(queries[i as usize].0 - 1) as usize];
            }
        }
        res
    }
}

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");
    let m: usize = input.trim().parse().expect("Invalid input");

    input.clear();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");
    let n: usize = input.trim().parse().expect("Invalid input");

    input.clear();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");
    let arr: Vec<i32> = input
        .trim()
        .split_whitespace()
        .map(|s| s.parse().expect("Invalid input"))
        .collect();

    input.clear();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");
    let queries: Vec<(i32, i32)> = input
        .trim()
        .split_whitespace()
        .map(|s| {
            let mut iter = s.split(',');
            (
                iter.next().unwrap().parse().expect("Invalid input"),
                iter.next().unwrap().parse().expect("Invalid input"),
            )
        })
        .collect();
    let res = DSA::range_sum_queries(m, n, arr.clone(), queries.clone());
    println!("{:?}", arr);
    for i in 0..n {
        println!("{:?} - {}", queries[i], res[i]);
    }
}
